packages/stac/lib/src/framework/stac_app.dart

```diff
@@ -224,4 +245,76 @@ class StacApp extends StatelessWidget {
       scrollBehavior: scrollBehavior,
     );
   }
+
+  Future<_ResolvedStacThemes> _resolveThemes() {
+    final themeInput = theme;
+    final darkThemeInput = darkTheme;
+
+    // Both themes are optional, so we need to handle null cases
+    final Future<StacTheme?>? themeFuture = themeInput?.resolve();
+    final Future<StacTheme?>? darkThemeFuture = darkThemeInput?.resolve();
+
+    // If both are null, return immediately with null themes
+    if (themeFuture == null && darkThemeFuture == null) {
+      return Future.value(_ResolvedStacThemes(theme: null, darkTheme: null));
+    }
+
+    return Future<_ResolvedStacThemes>(() async {
+      final resolvedTheme =
+          await (themeFuture ?? Future<StacTheme?>.value(null));
+      final resolvedDarkTheme =
+          await (darkThemeFuture ?? Future<StacTheme?>.value(null));
+
+      return _ResolvedStacThemes(
+        theme: resolvedTheme,
+        darkTheme: resolvedDarkTheme,
+      );
+    });
+  }
+
+  Widget _withResolvedThemes(
+    BuildContext context,
+    Widget Function(BuildContext, _ResolvedStacThemes) builder,
+  ) {
+    final resolved = _resolveThemes();
+    return FutureBuilder<_ResolvedStacThemes>(
+      future: resolved,
+      builder: (futureContext, snapshot) {
+        if (snapshot.connectionState == ConnectionState.waiting) {
+          return const _ThemeFutureLoading();
+        }
+        if (snapshot.hasError) {
+          Log.w('Failed to resolve theme: ${snapshot.error}');
+          return builder(
+            futureContext,
+            _ResolvedStacThemes(theme: null, darkTheme: null),
+          );
+        }
+        final themes = snapshot.data;
+        if (themes == null) {
+          return builder(
+            futureContext,
+            _ResolvedStacThemes(theme: null, darkTheme: null),
+          );
+        }
+        return builder(futureContext, themes);
+      },
+    );
+  }
+}
+
+class _ResolvedStacThemes {
+  const _ResolvedStacThemes({required this.theme, required this.darkTheme});
+
+  final StacTheme? theme;
+  final StacTheme? darkTheme;
+}
+
```

Comment:
Can we make this as a record?
`typedef ResolvedStacThemes = ({StacTheme? theme, StacTheme? darkTheme});`